Library/Homebrew/sbom.rb

```diff
@@ -0,0 +1,354 @@
+      "loaded_from_api"         => false,
+      "time"                    => Time.now.to_i,
+      "source_modified_time"    => formula.source_modified_time.to_i,
+      "compiler"                => compiler,
+      "stdlib"                  => stdlib,
+      "aliases"                 => formula.aliases,
+      "runtime_dependencies"    => SBOM.runtime_deps_hash(formula, runtime_deps),
+      "arch"                    => Hardware::CPU.arch,
+      "license"                 => SPDX.license_expression_to_string(formula.license),
+      "built_on"                => DevelopmentTools.build_system_info,
+      "source"                  => {
+        "path"         => formula.specified_path.to_s,
+        "tap"          => formula.tap&.name,
+        "tap_git_head" => nil, # Filled in later if possible
+        "spec"         => formula.active_spec_sym.to_s,
+        "patches"      => formula.stable&.patches,
+        "bottle"       => formula.bottle_hash,
+        "stable"       => {
+          "version"  => formula.stable&.version,
+          "url"      => formula.stable&.url,
+          "checksum" => formula.stable&.checksum,
+        },
+      },
+    }
+
+    # We can only get `tap_git_head` if the tap is installed locally
+    attributes["source"]["tap_git_head"] = T.must(formula.tap).git_head if formula.tap&.installed?
+
+    new(attributes)
+  end
+
+  sig { params(attributes: Hash).void }
+  def initialize(attributes = {})
+    attributes.each { |key, value| instance_variable_set(:"@#{key}", value) }
+  end
+
+  sig { returns(T::Boolean) }
+  def valid?
+    data = to_spdx_sbom
+
+    schema_string, _, status = Utils::Curl.curl_output(SCHEMA)
+
+    opoo "Failed to fetch schema!" unless status.success?
+
+    Homebrew.install_bundler_gems!(groups: ["bottle"])
+    require "json_schemer"
+
+    schemer = JSONSchemer.schema(schema_string)
+
+    return true if schemer.valid?(data)
+
+    opoo "SBOM is not valid, not writing to disk!"
+    schemer.validate(data).to_a.each do |error|
+      odebug error["error"]
+    end
+
+    false
+  end
+
+  sig { void }
+  def write
+    # If this is a new installation, the cache of installed formulae
+    # will no longer be valid.
+    Formula.clear_cache unless spdxfile.exist?
+
+    self.class.cache[spdxfile] = self
+
+    return unless valid?
+
+    spdxfile.atomic_write(JSON.pretty_generate(to_spdx_sbom))
+  end
+
+  sig { params(runtime_dependency_declaration: T::Array[Hash], compiler_declaration: Hash).returns(T::Array[Hash]) }
+  def generate_relations_json(runtime_dependency_declaration, compiler_declaration)
+    runtime = runtime_dependency_declaration.map do |dependency|
+      {
+        "spdxElementId"      => dependency["SPDXID"],
+        "relationshipType"   => "RUNTIME_DEPENDENCY_OF",
+        "relatedSpdxElement" => "SPDXRef-Bottle-#{name}",
+      }
+    end
+    patches = source["patches"].map do |_patch|
+      {
+        "spdxElementId"      => "SPDXRef-Patch-#{name}",
```

Comment:
Would be good to have tests cover this.

Reply:
Not sure how to do that without any source. Do we test those somewhere already?

Reply:
Unsure what "source" means here? Source code, formula source, something else?

Reply:
Well, to add patches to the formula I need to have something to patch. But I don't know how to do that in a test where the url is "foo-1.0"

Reply:
Yeh, there are other tests that use more wide-ranging Formula DSL features like this.